### PR TITLE
[AMD][Perf] Fuse GatedDeltaNet QKVZBA split/reshape/cat into a single Triton kernel for Qwen3.5-architecture MoE on HIP

### PR DESCRIPTION
**Pull request by yichiche**
## Motivation

`Qwen3_5GatedDeltaNet.forward` only takes the fused Triton split/reshape/cat path when the head-group
ratio `num_v_heads // num_k_heads` is 1, 2 or 4. Layouts with a ratio of 8 fall back to the eager
sequence (`fix_query_key_value_ordering` -> `.contiguous()` -> `torch.cat`), which costs four separate
data-movement kernels per GDN layer. Measured on a decode CUDA graph on MI355X (TP=8, aiter backend):

| # | Kernel | Source expression | us |
|---|---|---|---:|
| 1 | `direct_copy_kernel_cuda` (elementwise) | `b.contiguous()` | 3.8 |
| 2 | `direct_copy_kernel_cuda` (elementwise) | `a.contiguous()` | 4.0 |
| 3 | `CatArrayBatchedCopy` | `torch.cat((q, k, v))` -> `mixed_qkv` | 4.8 |
| 4 | `direct_copy_kernel_cuda` (elementwise) | materialising the non-contiguous `z` for `z.reshape(...)` | 4.4 |
| | **Total** | | **17.0** |

That is 17.0 us and 4 launches of pure data movement per GatedDeltaNet layer. Summed over the
GatedDeltaNet layers of the model measured here it accounts for ~1.17 ms of every decode iteration,
and the whole-trace `cat`/`contiguous`/`reshape` bucket was the 6th-largest kernel category at 4.9%
of GPU time.

`fused_qkvzba_split_reshape_cat_contiguous` already performs exactly this work in a single pass, and
it also writes `z` out contiguous so the later `z.reshape(...)` becomes free — i.e. it subsumes all
four kernels, not just the three obvious ones. It was simply never enabled for ratio 8, and its
hard-coded `num_warps=1` is too narrow once each program has to move 8 v-heads' worth of elements.

This PR enables the fused path for ratio 8 on aiter and scales `num_warps` to the per-program work,
replacing 4 kernels with 1 (3 launches eliminated per layer, 207 per decode iteration) and cutting
the replaced region by **-73.5%**.

## Modifications

- **python/sglang/kernels/ops/attention/triton_gdn_fused_proj.py**: in
  `fused_qkvzba_split_reshape_cat_contiguous`, replace the hard-coded `num_warps=1` with a heuristic
  based on how many v/z elements a single program moves
  (`v_elems_per_program = (num_heads_v // num_heads_qk) * head_v`): one warp up to 512 elements, four
  warps above that, so a wide head group does not serialise its vector loads/stores on a single warp.
  For every ratio the path already enabled at today's head dims the heuristic resolves back to
  `num_warps=1`, so existing callers keep the identical launch configuration. The kernel body itself
  is unchanged.

- **python/sglang/srt/models/qwen3_5.py**: hoist the fused-path ratio whitelist into a module-level
  `_GDN_FUSED_QKVZBA_RATIOS`, which is `(1, 2, 4, 8)` when `_use_aiter` is set and `(1, 2, 4)`
  otherwise, and use it in the `Qwen3_5GatedDeltaNet.forward` dispatch instead of the inline
  `in [1, 2, 4]` test. Because the non-aiter tuple is exactly the previous list, the CUDA / CPU / NPU
  control flow is unchanged — same branch taken, same kernels, byte-identical behaviour. The eager
  fallback is retained untouched for every ratio not in the tuple.

## Accuracy Tests

Model: a large MoE model with GatedDeltaNet linear-attention layers (MXFP4 weights, TP=8, MI355X),
aiter backend, ratio-8 head groups (`num_k_heads_tp=2`, `num_v_heads_tp=16`).

| Benchmark | Score | Threshold | Status |
|-----------|:-----:|:---------:|:------:|
| GSM8K (1319 questions, parallel=512, max-tokens 2048) | 0.9795 | 0.900 | PASS |

Run via `sglang.test.run_eval` against a live server. The baseline (fused path disabled, eager
fallback) scored 0.9787 on the same checkpoint and the same harness, so the fused kernel is
numerically equivalent to within run-to-run sampling noise.

## Speed Tests and Profiling

Workload: canonical-8k, IL=8192 / OL=1024, `num_prompts = concurrency x 10`. Baseline is the same
build with the change reverted. Traces captured on TP rank 0 at both anchors (conc4, conc64).

### Kernel-level (per GatedDeltaNet layer, decode)

| Kernel | Before (us) | After (us) | Notes |
|--------|:-:|:-:|-------|
| `direct_copy_kernel_cuda` (`b.contiguous()`) | 3.8 | -- | eliminated |
| `direct_copy_kernel_cuda` (`a.contiguous()`) | 4.0 | -- | eliminated |
| `CatArrayBatchedCopy` (`torch.cat((q,k,v))`) | 4.8 | -- | eliminated |
| `direct_copy_kernel_cuda` (`z` materialisation for `z.reshape`) | 4.4 | -- | eliminated; `z` now written contiguous |
| `fused_qkvzba_split_reshape_cat_contiguous_kernel` | -- | 4.5 | new fused kernel, 1 launch |
| **Replaced region total** | **17.0** | **4.5** | **-12.5 us (-73.5%), 4 kernels -> 1** |

Summed over every GatedDeltaNet layer of the model, that 12.5 us/layer is **862.5 us saved per decode
iteration** (region total 1173.0 us -> 310.5 us). The full-attention layers are untouched (0 us). At
conc64 the same region is 18.0 us -> 4.5 us per layer, i.e. 1242.0 us -> 310.5 us (**-75.0%**),
931.5 us saved per iteration. Kernel count in the decode GDN layer drops 13 -> 10 and its
kernel-time sum 86.5 us -> 75.3 us (conc4); 117.2 us -> 102.4 us (conc64).

Prefill benefits too, since the same dispatch is used there: the GDN region goes 60.4 us -> 41.9 us
per layer at conc4 prefill (7112 tokens) and 63.0 us -> 41.7 us at conc64 chunked prefill
(7316 tokens).

Profiling confirmation (trace analysis, both anchors): the `Qwen3_5GatedDeltaNet` module has **zero**
`direct_copy_kernel_cuda` and **zero** `CatArrayBatchedCopy` launches remaining in the after traces
(the only surviving `direct_copy` launches belong to `LogitsProcessor`), and
`fused_qkvzba_split_reshape_cat_contiguous_kernel` appears with exactly one launch per GDN
invocation. Normalised per GDN invocation over the whole trace, the replaced kernels go
17.36 us -> 4.92 us at conc4 (-71.7%) and 69.82 us -> 13.07 us at conc64 (-81.3%). The intended
`num_warps` branch is the one taken: `num_v_heads_tp / num_k_heads_tp = 8`, `head_v = 128`
-> 1024 elements per program -> `num_warps=4`.

### E2E benchmark

**Concurrency 4** (baseline n=2, after n=2; run-to-run spread < 0.2% on both sides)

| Metric | Before | After | Delta |
|--------|:-:|:-:|:-:|
| Total throughput (tok/s) | 1910.86 | 1983.34 | +3.79% |
| Output throughput (tok/s) | 212.57 | 220.63 | +3.79% |
| Median ITL (ms) | 16.605 | 16.005 | -3.61% |
| Median TPOT (ms) | 17.855 | 17.240 | -3.44% |
| Median E2E latency (ms) | 16678.65 | 16106.44 | -3.43% |
| Median TTFT (ms) | 449.72 | 446.93 | -0.62% |

**Concurrency 64** (baseline n=2, after n=3)

| Metric | Before | After | Delta |
|--------|:-:|:-:|:-:|
| Total throughput (tok/s) | 10008.69 | 10131.36 | +1.23% |
| Output throughput (tok/s) | 1110.38 | 1123.99 | +1.23% |
| Median ITL (ms) | 30.335 | 29.590 | -2.46% |
| Median TPOT (ms) | 56.47 | 55.77 | -1.24% |
| Median E2E latency (ms) | 52773.70 | 52141.40 | -1.20% |
| Median TTFT (ms) | 544.98 | 499.14 | -8.41% |

### Kernel-to-E2E consistency

| Anchor | Saving / iteration | Baseline iteration | Predicted ITL | Observed ITL |
|---|:-:|:-:|:-:|:-:|
| conc4 | 862.5 us | 16.605 ms | -5.19% | -3.61% |
| conc64 | 931.5 us | 30.335 ms | -3.07% | -2.46% |

Observed lands at ~70-80% of the GPU-side prediction at both anchors — same direction, same order of
magnitude — with the residual being the host/scheduler overhead floor that a GPU-side fusion does not
remove. Both concurrencies were repeated on both sides and the run spread (<0.2% at conc4, 0.13% for
the conc64 baseline) is far smaller than the reported deltas. TTFT is not where this change pays off:
the ~1.28 ms saved across a whole prefill is ~0.3% of the baseline TTFT, so the TTFT deltas above are
dominated by queueing rather than by the fused kernel.

### Notes

- The `num_warps` threshold (512 v/z elements per program) is a static heuristic tuned on MI355X for
  `head_v=128` rather than an autotuned config; ratio-8 enablement stays behind `_use_aiter` so other
  backends are unaffected until they are measured.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #31654038211](https://github.com/sgl-project/sglang/actions/runs/31654038211)
Latest PR Test (Extra): :x: [Run #31654038134](https://github.com/sgl-project/sglang/actions/runs/31654038134)

### CODE REVIEW

File: python/sglang/kernels/ops/attention/triton_gdn_fused_proj.py [hunk 1]
```diff
@@ -293,6 +293,12 @@ def fused_qkvzba_split_reshape_cat_contiguous(
     )
     a = torch.empty_like(b)
     grid = (batch * seq_len, num_heads_qk)
+    # Each program moves `v_per_group * head_v` elements for both v and z. For
+    # the small head-group ratios (<= 512 elements) a single warp is the best
+    # fit; wider ratios (e.g. 8 v-heads per k-head) need more lanes so the
+    # per-program vector load/store does not serialize.
+    v_elems_per_program = (num_heads_v // num_heads_qk) * head_v
+    num_warps = 1 if v_elems_per_program <= 512 else 4
```
Comment (HaiShaw) [inline python/sglang/kernels/ops/attention/triton_gdn_fused_proj.py]: let's confine this changes to hip/aiter only.

Comment (yichiche) [reply python/sglang/kernels/ops/attention/triton_gdn_fused_proj.py]: Add if_hip guard, kick-off ci again.

**Review by HaiShaw — APPROVED** [review]
